File: android/src/main/scala/sgl/android/GameView.scala
package sgl
package android

import _root_.android.content.Context
import _root_.android.app.Activity
import _root_.android.content.res.Resources
import _root_.android.graphics.Canvas
import _root_.android.graphics.Color
import _root_.android.graphics.Paint
import _root_.android.graphics.Paint.Style
import _root_.android.util.AttributeSet
import _root_.android.view.View
import _root_.android.view.SurfaceView
import _root_.android.view.SurfaceHolder;
import _root_.android.widget.LinearLayout
import _root_.android.graphics.Bitmap;
import _root_.android.graphics.BitmapFactory;
import _root_.android.graphics.PorterDuff

import _root_.android.app.AlertDialog
import _root_.android.content.DialogInterface

class GameView(
  val gameActivity: AndroidApp, 
  attributeSet: AttributeSet
) extends SurfaceView(gameActivity) with SurfaceHolder.Callback { self =>

  private implicit val LogTag = gameActivity.Logger.Tag("sgl-gameview")

  getHolder.addCallback(this)
  setFocusable(true)

  var surfaceValid: Boolean = false

  override def surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int): Unit = {
    gameActivity.logger.debug("SurfaceChanged called")
  }

  override def surfaceCreated(holder: SurfaceHolder): Unit = {
    gameActivity.logger.debug("SurfaceCreated called")
    surfaceValid = true
    if(gameActivity.isRunning)
      gameActivity.resume()
  }

  override def surfaceDestroyed(holder: SurfaceHolder): Unit = {
    gameActivity.logger.debug("SurfaceDestroyed called")
    surfaceValid = false
    if(gameActivity.isRunning) {
      gameActivity.pause()
      gameActivity.gameLoop.runningThread.join
    }
  }

  def this(gameActivity: AndroidApp) = this(gameActivity, null)

}

File: android-kotlin/app/src/main/java/sgl/android/GameView.kt
package sgl.android

import android.util.AttributeSet
import android.view.SurfaceView
import android.view.SurfaceHolder

class GameView(val app: BaseMainActivity, attributeSet: AttributeSet?): SurfaceView(app), SurfaceHolder.Callback {

    //private implicit val LogTag = Logger.Tag("sgl-gameview")

    init {
        holder.addCallback(this)
        this.setFocusable(true)
        this.setContentDescription("Main View where the game is rendered.")
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int): Unit {
        //logger.debug("SurfaceChanged called")
    }

    override fun surfaceCreated(holder: SurfaceHolder): Unit {
        //logger.debug("SurfaceCreated called")
        app.surfaceReady = true
    }

    override fun surfaceDestroyed(holder: SurfaceHolder): Unit {
        //logger.debug("SurfaceDestroyed called")
        app.surfaceReady = false

        // We must ensure that the thread that is accessing this surface
        // does no longer run at the end of the surfaceDestroyed callback,
        // so we join on the thread.

        // this is safe to call twice (also called in onPause)
        app.gameLoop?.running = false
        // The thread should finish the current frame and then return the run method,
        // so we just join on the thread to make sure there are no more rendering calls
        // after returning this callback.
        app.gameLoopThread?.join()
    }

}